synapse/rest/client/v1/login.py

```diff
@@ -134,6 +142,24 @@ async def on_POST(self, request: SynapseRequest):
             result["well_known"] = well_known_data
         return 200, result
 
+    async def _do_appservice_login(
+        self, login_submission: JsonDict, appservice: ApplicationService
+    ):
+        logger.info(
+            "Got appservice login request with identifier: %r",
+            login_submission.get("identifier"),
+        )
+        identifier = convert_client_dict_legacy_fields_to_identifier(login_submission)
```

Comment:
I think you want to check `identifier["type"]` after conversion to ensure it is `m.uid.user`.

It might be worth abstracting lines 246 - 256 to a `_get_qualified_user_id` method.

Reply:
done